### PR TITLE
feat: switch to b+ tree

### PR DESCRIPTION
**Pull request by kevmo314**

### CODE REVIEW

File: pkg/handlers/jsonl.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+			fieldOffset := dec.InputOffset() + 1 // skip the :
+
+			value, err := dec.Token()
+			if err != nil {
+				return fmt.Errorf("failed to read token: %w", err)
+			}
+
+			name := strings.Join(append(path, key), ".")
+
+			if name != "VendorID" {
```
Comment (friendlymatthew) [inline pkg/handlers/jsonl.go:119]: I assume this logic is for a bug?

Comment (kevmo314) [reply pkg/handlers/jsonl.go:119]: Yeah, it's some debugging code to repro

File: pkg/appendable/index_file.go [hunk 2]
```diff
@@ -1,72 +1,186 @@
...
+	return !exists, nil
 }
 
-func (i *IndexFile) findIndex(name string, value any) int {
-	// find the index for the field
-	match := -1
-	for j := range i.Indexes {
-		if i.Indexes[j].FieldName == name {
-			match = j
+func (i *IndexFile) IndexFieldNames() ([]string, error) {
```
Comment (friendlymatthew) [inline pkg/appendable/index_file.go:100]: this function iterates through meta pages and retrieves all field names

File: pkg/handlers/jsonl_test.go [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+		if err := md1.UnmarshalBinary(buf1); err != nil {
+			t.Fatal(err)
+		}
+		if md1.FieldType != appendable.FieldTypeString {
+			t.Errorf("got i.Indexes[0].FieldType = %#v, want FieldTypeString", md1.FieldType)
+		}
+
+		v2 := make([]byte, 8)
+		binary.LittleEndian.PutUint64(v2, math.Float64bits(123))
+		mp2, found, err := collected[1].BPTree(r2).Find(v2)
```
Comment (friendlymatthew) [inline pkg/handlers/jsonl_test.go:244]: It's unable to find `123` from the bptree. However, we can assert the `Find()` is not the issue as the `csv_test.go` has a test case that passes this assertion.

I don't see any glaring differences in the logic. The debugger indicates both values are in the bptree:

```sh
=== RUN   TestJSONL/existing_index_but_different_type
time=2024-02-02T10:28:43.780-05:00 level=DEBUG msg="handling json token" value=test1 start=8
time=2024-02-02T10:28:43.780-05:00 level=INFO msg="read line" i=0 offset=17
time=2024-02-02T10:28:43.780-05:00 level=DEBUG msg="handling json token" value=123 start=8
time=2024-02-02T10:28:43.780-05:00 level=INFO msg="read line" i=0 offset=30
```

File: pkg/btree/pagefile.go [hunk 4]
```diff
@@ -65,65 +73,107 @@ func (pf *PageFile) Page(i int) (int64, error) {
...
-			if _, err := pf.ReadWriteSeeker.Seek(int64(j*8), io.SeekStart); err != nil {
-				return err
-			}
-			return nil
+	if pf.freePageCount == len(pf.freePageIndexes) {
+		return errors.New("free page index is full")
+	}
+
+	for i, _ := range pf.freePageIndexes {
+		if pf.freePageIndexes[i] == offset {
```
Comment (friendlymatthew) [inline pkg/btree/pagefile.go:166]: extra test case to check if the offset already exists within the free pages

**Pull request merged**